keps/sig-node/5532-restart-pod-on-container-exits/README.md

```diff
@@ -0,0 +1,822 @@
+
+Another scenario is when Init container "takes the next item from the queue". And main container exists with the indication that it want's a "new item' to process. See also
+
+- https://github.com/kubernetes/enhancements/issues/3676
+- https://github.com/kubernetes/enhancements/issues/3759#issuecomment-2389506153
+
+**Handles Init Container Failures**
+
+Sidecar container failures sometimes render the main container not ready, and restarting the sidecar is insufficient. For example, if a sidecar that manages the remote volume fails and restarts, the main container may be trying to access an outdated volume. With RestartPod action, the sidecar could force the main container to restart as well for a clean environment.
+
+**Efficient In-Place Restart**
+
+Deleting and recreating a pod is a heavy operation involving the scheduler, node resource allocation, and re-initialization of networking and storage. An in-place restart, which preserves the pod sandbox and its associated resources (UID, IP, devices), is significantly faster and reduces resource churn. 
+
+This is especially helpful for ML training workloads, where computation resources are expensive and in-place restarts improves resource usage efficiency. This is also helpful if the workload can run in seconds, and restart in-place is much more efficient than rescheduling. See also
+
+- https://docs.google.com/document/d/16zexVooHKPc80F4dVtUjDYK9DOpkVPRNfSv0zRtfFpk/edit?tab=t.0#heading=h.y6xl7juq7465
+
+**Separates Watcher-sidecars from worker containers**
+
+In ML training workloads we have setups with a watcher process that listens for failures, and restarts the worker from the previous checkpoint if needed. Without a RestartPod action, the watcher process and worker process have to be coupled in a single container, increasing complexity and decreasing cohesion. The RestartPod action eliminates this coupling. See also
+
+- https://docs.google.com/document/d/16zexVooHKPc80F4dVtUjDYK9DOpkVPRNfSv0zRtfFpk/edit?tab=t.0#heading=h.y6xl7juq7465
+- https://github.com/kubernetes/enhancements/issues/4438 
+
+**Improved Predictability and Debugging**
+
+Restarting all containers together brings the entire pod to a known good state. This is often easier to reason about and debug than a state where some containers are running while others have been restarted independently.
+
+### Goals
+
+- Introduce a `RestartPod` action to the `ContainerRestartRule` API.
+- Implement the kubelet logic to perform an in-place pod restart, which includes:
+<<[UNRESOLVED whether containers should be killed or terminated gracefully]>>
+    - Terminating all containers (not including graceful termination).
+<<[/UNRESOLVED]>>
+    - Re-running init containers.
+    - Restarting all regular and sidecar containers.
+    - Preserving the pod sandbox, UID, and network identity.
+- Introduce a new PodCondition to make the pod restart process observable.
+
```

Comment:
let's introduce some perf goals here. Perhaps "under 1 minute since CRI detected the termination"